### ISSUE TITLE
Indexed properties. throwing exception while queirying.

### ISSUE THREAD

**Issue by KirubakaranSP** [hugegraph/hugegraph#1693]
### Problem Type (问题类型)

gremlin (结果不合预期)

### Before submit

- [X] 我已经确认现有的 [Issues](https://github.com/hugegraph/hugegraph/issues) 与 [FAQ](https://hugegraph.github.io/hugegraph-doc/guides/faq.html) 中没有相同 / 重复问题

### Environment (环境信息)

- Server Version: v0.11.x
- Backend: RocksDB x nodes, HDD or SSD 
- OS: xx CPUs, xx G RAM, Centos 7.x 
- Data Size:  xx vertices, xx edges 

### Your Question (问题描述)

I've indexed my vertex properties (search index and secondary index).But while querying it throws error .

Exception : Exception ::Don't accept query based on properties [NAME, ADDRESS] that are not indexed in label 'DETAILS', may not match secondary condition

But the above mentioned properties in vertex DETAILS have been indexed.

"NAME" ----> Search Index
"ADDRESS" ----> Secondary Index

NOTE : I've previously used the same configurations and worked perfectly .I've recently used new hugegraph build ,So I am wondering is that creating any problem with the configurations in hugegraph.properties.

Configurations :

Hugegraph version : 0.11.2
backend=rocksdb
serializer=binary
store=hugegraph

search.text_analyzer=jieba
search.text_analyzer_mode=INDEX (tried changing these two properties with ikanalyzer )

**Comment by KirubakaranSP** [comment]
Any update on this ?

**Comment by javeme** [comment]
@KirubakaranSP Could you post the query statement and possible error log?

**Comment by KirubakaranSP** [comment]
Vertex Label : **DETAILS**
Properties : 
**NAME - > Search index
ADDRESS - >Standard Index**

Query : 
g.V().has('DETAILS','NAME','Tom').has('DETAILS','ADDRESS','Wellington Street')

Exception : 
ERROR :: com.baidu.hugegraph.exception.NoIndexException: Don't accept query based on properties [NAME, ADDRESS] that are not indexed in label 'DETAILS', may not match secondary condition

**Comment by KirubakaranSP** [comment]
Any update on this ??

**Comment by javeme** [comment]
Could you try query like this: `g.V().has('DETAILS','NAME',Text.contains('Tom')).has('DETAILS','ADDRESS','Wellington Street')`

**Comment by KirubakaranSP** [comment]
Can you please share what are the Text class methods available other than contains?

**Comment by javeme** [comment]
There are no other methods in Text class.

**Comment by KirubakaranSP** [comment]
Is **Text.contains('Tom')** only check part of the string that contains the word Tom or the whole string?
If it checks only part of the string ,then which method has to be used to check the string and vice versa?

**Comment by javeme** [comment]
There are three syntaxes:

1. `Text.contains('Tom Cat')`: means "Tom" or "Cat"
2. `Text.contains('(Tom Cat)')`: means "Tom Cat"
3. `Text.contains('(Tom Cat | Jerry)')`: means "Tom Cat" or "Jerry"

You can use the first or the third syntax.

**Comment by KirubakaranSP** [comment]
Thanks lot

**Comment by KirubakaranSP** [comment]
Is there any issues in 0.12.0 version of hugegraph? It takes lot of time to execute a query than 0.11.2 version.

**Comment by KirubakaranSP** [comment]
How to match the whole string which are indexed using full-text index?I've tried the solution you mentioned above but it is not working.

Query ::  g.V().has('DETAILS', 'NAME', Text.contains('(Tom)'))

I want to return vertices having name 'Tom' but it is not working.

**Comment by javeme** [comment]
Did you modify the source code? https://github.com/hugegraph/hugegraph/issues/1638#issuecomment-993562786 (It can't match the whole string If the full text is not added to segments)

**Comment by KirubakaranSP** [comment]
Did u mean to uncomment the below line ?

 ```
 /*
         * Add original text to segments at the insertion stage,
         * in order to can match fully words at the query stage.
         */
        // segments.add(text);
```

**Comment by javeme** [comment]
You are right. that behavior is expected, because the `Text.contains('JERRY')` operation is doing a filtering (not querying from index table).

If we need a filtering by the meaning of `Text.contains('(JERRY)')`, just keep statement:
 `has('DETAILS', 'NAME', 'JERRY')`.

Or we can adapt the code if want to directly use `Text.contains('(JERRY)')`:
https://github.com/hugegraph/hugegraph/blob/43b8ae7bff7cdec5a8659af06eedc56307de4d2c/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Condition.java#L77

**Comment by KirubakaranSP** [comment]
How can we query like has('DETAILS', 'NAME', 'JERRY') when the attribute is search indexed.It will throw exception right?
Exception :
Gremlin execute failed, the details: Don't accept query based on properties [NAME] that are not indexed in label'DETAILS', may not match secondary condition

**Comment by javeme** [comment]
can you paste the full query statement?

please try:
`g.V().has('DETAILS', 'NAME', Text.contains('TOM')).has('DETAILS', 'ADDRESS', 'WT Street').repeat(outE().inV().simplePath()).until(and(has('DETAILS', 'NAME', 'JERRY').has('DETAILS', 'ADDRESS', 'Mathew Street'))).path()`

**Comment by KirubakaranSP** [comment]
Yea tha above syntax works.Can you explain how it works,as both of the attributes were same ?

**Comment by KirubakaranSP** [comment]
When i add  `segments.add(text);` to the code then constant values having special characters throwing the error we discussed [here](https://github.com/hugegraph/hugegraph/issues/1688) .Did u check this behaviour before ?

**Comment by KirubakaranSP** [comment]
any update on this?

**Comment by javeme** [comment]
It behaves as expected, you need to escape the text before `segments.add(text)`

**Comment by KirubakaranSP** [comment]
Is the below code right?
` private Set<String> segmentWords(String text) {
        /*
         Support 3 kinds of query:
         - Text.contains("(word)"): query by user-specified word;
         - Text.contains("(word1|word2|word3)"): query by user-specified words;
         - Text.contains("words"): query by words splitted from analyzer;
         Note: all kinds support words exact match
         */
        if (text.startsWith(START_SYMBOL) && text.endsWith(END_SYMBOL)) {
            String subText = text.substring(1, text.length() - 1);
            if (subText.contains(WORD_DELIMITER)) {
                String[] texts = StringUtils.split(subText, WORD_DELIMITER);
                return ImmutableSet.copyOf(texts);
            } else {
                return ImmutableSet.of(subText);
            }
        }
        
        Set<String> segments = this.textAnalyzer.segment(text);

        text = text.replace("\u0000", "\u0003");
        text= '\u0003' + text;
        /*
         * Add original text to segments at the insertion stage,
         * in order to can match fully words at the query stage.
         */
         segments.add(text);

        // Ignore unicode \u0000 to \u0003
        segments.removeAll(ConditionQuery.IGNORE_SYM_SET);
        return segments;
    }`

Here are my usecase?
1.Perform full word search for vertices indexed using full-text search?(Example cases discussed above).
2.Ignore special characters in constant value.

**Comment by javeme** [comment]
Yes, it looks right.

**Comment by KirubakaranSP** [comment]
But after adding this full- text search is not working?

**Comment by javeme** [comment]
> But after adding this full- text search is not working?

Could you show more details about query statement and error message? 

can remove the line `text= '\u0003' + text` and add some log at that line.

**Comment by KirubakaranSP** [comment]
here is the query
`g.V().has('DETAILS', 'NAME', Text.contains('(TOM)')).has('DETAILS', 'ADDRESS', 'WT Street').repeat(outE().inV().simplePath()).until(and(has('DETAILS', 'NAME', Text.contains('(JERRY)')).has('DETAILS', 'ADDRESS', 'Mathew Street'))).path()

`

**Comment by javeme** [comment]
> You are right. that behavior is expected, because the `Text.contains('JERRY')` operation is doing a filtering (not querying from index table).
> 
> If we need a filtering by the meaning of `Text.contains('(JERRY)')`, just keep statement: `has('DETAILS', 'NAME', 'JERRY')`.
> 
> Or we can adapt the code if want to directly use `Text.contains('(JERRY)')`:
> 
> https://github.com/hugegraph/hugegraph/blob/43b8ae7bff7cdec5a8659af06eedc56307de4d2c/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Condition.java#L77

@KirubakaranSP Hi, Could you please contribute a PR to adapt filtering by `Text.contains('(JERRY)')`?
